smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/HttpChecksumTraitValidator.java

```diff
@@ -0,0 +1,173 @@
+            if (property.getLocations().isEmpty()) {
+                events.add(error(operation, trait,
+                        "The `request` property of the `httpChecksum` trait must contain at least one `location` "
+                                + "property entry, found none."));
+            }
+            if (property.getAlgorithms().isEmpty()) {
+                events.add(error(operation, trait,
+                        "The `request` property of the `httpChecksum` trait must contain at least one `algorithms` "
+                                + "property entry, found none."));
+            }
+
+            if (operation.getInput().isPresent()) {
+                Shape shape = model.expectShape(operation.getInput().get());
+                events.addAll(validatePrefix(operation, shape, property.getPrefix(), "request"));
+            }
+        });
+
+        responseProperty.ifPresent(property -> {
+            if (property.getLocations().isEmpty()) {
+                events.add(error(operation, trait,
+                        "The `response` property of the `httpChecksum` trait must contain at least one `location` "
+                                + "property entry, found none."));
+            }
+            if (property.getAlgorithms().isEmpty()) {
+                events.add(error(operation, trait,
+                        "The `response` property of the `httpChecksum` trait must contain at least one `algorithms` "
+                                + "property entry, found none."));
+            }
+
+            if (operation.getOutput().isPresent()) {
+               Shape shape = model.expectShape(operation.getOutput().get());
+               events.addAll(validatePrefix(operation, shape, property.getPrefix(), "response"));
+            }
+
+            if (!operation.getErrors().isEmpty()) {
+                for (ShapeId id: operation.getErrors()) {
+                   Shape shape = model.expectShape(id);
+                   events.addAll(validatePrefix(operation, shape, property.getPrefix(), "response"));
+                }
+            }
+        });
+
+        return events;
+    }
+
+
+    /**
+     * Validates if prefix used with the httpChecksum trait conflicts with headers modeled using the
+     * `httpPrefixHeaders` trait or `httpHeader` trait.
+     *
+     * @param operation is the operation shape on which httpChecksum trait is modeled
+     * @param containerShape is the shape referenced by the operation shape, this can be either input shape
+     *                       or output shape
+     * @param prefix is the prefix modeled on the httpChecksum trait
+     * @param propertyType is the property type being validated for httpChecksum trait, this can be either `request`
+     *                    or `response` property
+     * @return list of triggered validation events
+     */
+    private List<ValidationEvent> validatePrefix(
+            OperationShape operation, Shape containerShape,
```

Comment:
Having `StructureShape containerShape` (see above) means you don't need a `Walker` anymore, as you can just check the members of the structure shape for the potential conflicts.